pkg/kademlia/workers.go

```diff
@@ -0,0 +1,139 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package kademlia
+
+import (
+	"bytes"
+	"context"
+	"math/rand"
+	"sync"
+	"time"
+
+	"storj.io/storj/pkg/node"
+	proto "storj.io/storj/protos/overlay"
+)
+
+const (
+	uncontacted = iota
+	inProgress
+	completed
+)
+
+type chore struct {
+	status int
+	node   *proto.Node
+}
+
+type worker struct {
+	// node id to chore
+	contacted   map[string]*chore
+	mu          *sync.Mutex
+	maxResponse time.Duration
+	cancel      context.CancelFunc
+	nodeClient  node.Client
+	find        proto.Node
+	k           int
+}
+
+func (w *worker) work(ctx context.Context) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+			w.mu.Lock()
+			var wk *chore
+			for i, v := range w.contacted {
+				if v.status == uncontacted {
+					wk = v
+					w.contacted[i].status = inProgress
+					w.mu.Unlock()
+					break
+				}
+			}
+
+			if wk == nil {
+				w.mu.Unlock()
+				time.AfterFunc(2*w.maxResponse, w.cancel)
+				return nil
+			}
+
+			start := time.Now()
+			nodes, err := w.nodeClient.Lookup(ctx, *wk.node, w.find)
+			if err != nil {
+				return err
+			}
+			latency := time.Now().Sub(start)
+			if latency > w.maxResponse {
+				w.maxResponse = latency
+			}
+
+			w.mu.Lock()
+			if (len(w.contacted) + len(nodes)) < w.k {
+				for _, v := range nodes {
+					w.contacted[v.GetId()] = &chore{node: v, status: uncontacted}
+				}
+				w.mu.Unlock()
+			}
+
+			// sort all nodes from map
+			mapNodes := make([][]byte, len(w.contacted))
+			ii := 0
+			for i := range w.contacted {
+				mapNodes[ii] = []byte(i)
+				ii++
+			}
+
+			// sort all nodes from map
+			lookupNodes := make([][]byte, len(nodes))
+			for i, v := range nodes {
+				lookupNodes[i] = []byte(v.GetId())
+			}
+
+			self := []byte(w.find.GetId())
+			smapNodes := sortByXOR(mapNodes, self)
+			// sort all nodes returned from lookup
+			slookupNodes := sortByXOR(lookupNodes, self)
+			// update the nearest-k data structure with the results of the request, bumping nodes out of the data structure that are farther than k.
+			ll := 0
+			for i := len(smapNodes) - 1; i >= 0; {
+				if bytes.Compare(smapNodes[i], slookupNodes[ll]) < 0 {
+					i--
+					ll++
+				} else {
+					delete(w.contacted, string(smapNodes[i]))
+					w.contacted[string(slookupNodes[ll])] = &chore{node: nodes[ll], status: uncontacted}
+					i--
+					ll++
+				}
+			}
+
+			w.mu.Unlock()
+		}
+
+	}
+
+}
+
+// sortByXOR: helper, quick sorts node IDs by xor from local node, smallest xor to largest
```

Comment:
heh, i was telling jen this, but the new 1.8 sort.Slice might be way easier to use

Reply:
okay after reading the logic above, i think this goes away with the right implementation of container/heap